### ISSUE TITLE
Suggestion: Provide Option to use asyncio ProcessPool instead of ThreadPool

### ISSUE THREAD

**Issue by ericdrobinson** [Azure/azure-functions-python-worker#403]
One of the major use cases for Python is data science computation. This type of computation typically features _lots_ of complex, heavy math that may take a very long time to compute. This type of computation is _terrible_ for Python threads due to each thread sharing the [GIL](https://wiki.python.org/moin/GlobalInterpreterLock) and blocking all other threads. This is even mentioned in the `azure-functions-python-worker` [source code](https://github.com/Azure/azure-functions-python-worker/blob/b28328ebc813b1462e2efb7d92479dbc250ae1d4/azure/functions_worker/dispatcher.py#L49).

Unfortunately, Azure Functions in Python is [currently built](https://docs.microsoft.com/en-us/azure/azure-functions/functions-reference-python#async) to use an `asyncio` Thread Pool. What this means for compute-heavy functions is that the default concurrency settings for function instances are relatively meaningless for compute-heavy applications. The [HTTP trigger](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json#http)'s `maxConcurrentRequests` value defaults to 100. The _actual_ number of concurrent requests that can be handled with this approach? **1**. The [Queue Storage trigger](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-storage-queue#host-json)'s `batchSize` value defaults to 16. The _actual_ number of concurrent requests that can be handled with this approach? **1**.

Part of the attraction of Azure Functions is the simplicity of the programming model married with the promise of autoscale. That attraction disappears entirely when you get into the nitty-gritty and find that Azure Function _instances_ are supposed to handle multiple concurrent requests but your function can only handle one at a time due to Python's intrinsic limitations (as employed by the `azure-functions-python-worker`). Each Function App instance is run on its own CPU [as I understand it] which has (at time of writing) at least 2 logical cores. With CPU-bound processing tasks, it would certainly be nice to utilize that potential and get at least a _little_ concurrency back (say, 2 concurrent tasks).

Getting concurrency back (without resorting to subprocesses, which basically defeats the purpose of using the language in the first place) is _somewhat_ possible with the `asyncio` [`multiprocessing`](https://docs.python.org/3.6/library/multiprocessing.html) module. While using this module [does indeed work](https://github.com/Azure/azure-functions-python-worker/issues/236#issuecomment-470360431) (and enables some controllable measure of concurrency), it has some serious issues [with respect to testing](https://github.com/Microsoft/ptvsd/issues/1228#issuecomment-479959990). What's more, the default process start method uses `fork` under the hood, which has its own issues when run in an application [that has already spawned threads](https://www.evanjones.ca/fork-is-dangerous.html). This issue is even [called out](https://docs.python.org/3.6/library/multiprocessing.html#contexts-and-start-methods) in the Python documentation:

>Note that safely forking a multithreaded process is problematic.

A [suggested workaround](https://github.com/Microsoft/ptvsd/issues/1228#issuecomment-475030964) for these issues is to use the [_spawn_](https://docs.python.org/3.6/library/multiprocessing.html#contexts-and-start-methods) start method. Unfortunately, this [doesn't work](https://github.com/Microsoft/ptvsd/issues/1228#issuecomment-479959990) in Azure Functions with Python because we do not control the application startup process: by the time our function code is run it is too late to call [`set_start_method`](https://docs.python.org/3.6/library/multiprocessing.html#multiprocessing.set_start_method).

In summary, many desirable Python use cases are CPU-bound and cause significant issues/headaches when working with the current Azure Functions concurrency model. Changes (even if _optional_) to the default concurrency model used by the `azure-functions-python-worker` might significantly improve the ability of consumers to do more with each function instance; to better take advantage of Azure Functions platform and promise with Python.

**Comment by asavaritayal** [comment]
/cc @pragnagopa 

The work for this is already in progress. Closing as a duplicate of https://github.com/Azure/azure-functions-host/issues/4194. PR - https://github.com/Azure/azure-functions-host/pull/4210
